File: jun1/src/combinators/seq.py
from typing import list

from ..combinator import Combinator
from ..parse_iteration_result import FrameStack, ParserIterationResult
from ..state import CombinatorState


class SeqState(CombinatorState):
    def __init__(self, its: list[list[CombinatorState]]) -> None:
        self.its = its


class Seq(Combinator):
    def __init__(self, combinators: list[Combinator]) -> None:
        self.combinators = combinators

    def initial_state(self, signal_id: int, frame_stack: FrameStack) -> SeqState:
        return SeqState([
            [combinator.initial_state(signal_id, frame_stack.clone())]
            if i == 0 else []
            for i, combinator in enumerate(self.combinators)
        ])

    def next_state(self, state: SeqState, c: Optional[str], signal_id: int) -> tuple[SeqState, int, ParserIterationResult]:
        a_result = ParserIterationResult.new_empty()
        for i, (combinator, its) in enumerate(zip(self.combinators, state.its)):
            for j in range(len(its)):
                state.its[i][j], signal_id, result = combinator.next_state(state.its[i][j], c, signal_id)
                if i == 0:
                    a_result.merge_assign(result)
                else:
                    if a_result.is_complete:
                        b_it = combinator.initial_state(signal_id, a_result.frame_stack.clone())
                        b_it, signal_id, b_result = combinator.next_state(b_it, None, signal_id)
                        state.its[i].append(b_it)
                        a_result.forward_assign(b_result)
                    a_result.merge_assign(result)
        return state, signal_id, a_result

File: jun1/src/combinators/seq.rs
use std::rc::Rc;
use crate::combinator::Combinator;
use crate::helper_functions::{process, seq2_helper};
use crate::parse_iteration_result::{FrameStack, ParserIterationResult};
use crate::state::CombinatorState;

pub struct Seq(pub Rc<[Rc<dyn Combinator>]>);

impl Combinator for Seq {
    fn initial_state(&self, signal_id: &mut usize, frame_stack: FrameStack) -> Box<dyn CombinatorState> {
        let mut its = Vec::with_capacity(self.0.len());
        its.push(vec![self.0[0].initial_state(signal_id, frame_stack)]);
        for _ in 1..self.0.len() {
            its.push(Vec::new());
        }
        Box::new(SeqState { its })
    }

    fn next_state(&self, state: &mut dyn CombinatorState, c: Option<char>, signal_id: &mut usize) -> ParserIterationResult {
        let state = state.as_any_mut().downcast_mut::<SeqState>().expect("Invalid state type");
        let mut a_result = process(&self.0[0], c, &mut state.its[0], signal_id);
        for (combinator, its) in self.0.iter().zip(state.its.iter_mut()).skip(1) {
            let b_result = process(combinator, c, its, signal_id);
            seq2_helper(combinator, &mut a_result, b_result, its, signal_id);
        }
        a_result
    }
}

pub struct SeqState {
    pub its: Vec<Vec<Box<dyn CombinatorState>>>,
}

impl CombinatorState for SeqState {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }
}
